doc/src/build/programming-with-objects/ch5-dynamic-fields.md

```diff
@@ -0,0 +1,166 @@
+
+### Field Names
+
+Unlike an object's regular fields whose names must be Move identifiers, dynamic field names can be any value that has `copy`, `drop` and `store`. This includes all Move primitives (integers, booleans, byte strings), and structs whose contents all have `copy`, `drop` and `store`.
+
+### Adding Dynamic Fields
+
+Dynamic fields are added with the following APIs:
+
+```rust
+use sui::dynamic_field as field;
+use sui::dynamic_object_field as ofield;
+
+public fun field::add<Name: copy + drop + store, Value: store>(
+  object: &mut UID,
+  name: Name,
+  value: Value,
+)
+
+public fun ofield::add<Name: copy + drop + store, Value: key + store>(
+  object: &mut UID,
+  name: Name,
+  value: Value,
+);
+```
+
+These functions add a field with name `name` and value `value` to `object`. To see it in action, consider these code snippets:
+
+First we define two object types for the parent and the child:
+
+```rust
+struct Parent has key {
+    id: UID,
+}
+
+struct Child has key, store {
+    id: UID,
+}
+```
+
+Now, we can define an API to add a `Child` object as a dynamic field of a `Parent` object:
+
+```rust
+public entry fun add_child(parent: &mut Parent, child: Child) {
+    ofield::add(&mut parent.id, b"child", child);
+}
+```
+
+This function takes the `Child` object by value, and makes it a dynamic field of `parent` with name `b"child"` (a byte string). At the end of the `add_child` call, we have the following ownership relationship:
+
+1. Sender address (still) owns the `Parent` object.
+2. The `Parent` object owns the `Child` object, and can refer to it by the name `b"child"`.
+
+> :warning: It is an error to overwrite a field (attempt to add a field with the same Name type and value as one that is already defined), and a transaction that does this will abort.  Fields can be modified in-place by borrowing them mutably and can be overwritten safely (e.g. to change its value type) by removing the old value first (see below for details).
+
+### Accessing Dynamic Fields
+
+Dynamic fields can be accessed by reference using the following APIs:
+
+```rust
+use sui::dynamic_field as field;
+
+public fun borrow<Name: copy + drop + store, Value: store>(
+    object: &UID,
+    name: Name,
+): &Value
+
+public fun borrow_mut<Name: copy + drop + store, Value: store>(
+    object: &mut UID,
+    name: Name,
+): &mut Value
+```
+
+Where `object` is the UID of the object the field is defined on and `name` is the field's name.
+
+> :bulb: `sui::dynamic_object_field` has equivalent functions for object fields.
+
+Let's look at how to use these APIs with the `Parent` and `Child` types defined earlier:
+
+```rust
+public entry fun mutate_child(_child: &mut Child) {}
```

Comment:
Amusingly... this won't actually result in a mutation in storage :p Unless you call `write_ref`, the global value won't be dirtied and it will be invisible to the object runtime (at least as it is currently written). 
Might want to just bump a counter or something?